### PR TITLE
[Precision Depth Alignment] Add support for CUDNN to paddle.nn.functional.grid_sample to align with torch accuracy. 

### PR DESCRIPTION
**Pull request by zhengshengning**
### PR Category

Operator Mechanism

### PR Types

New features

#### 背景说明

在 `padding_mode == zeros` 且 `mode == bilinear` 且 `align_corners == true` 的配置下，Paddle 的原始实现与 PyTorch 存在一定数值差异。通过调用 cuDNN 提供的 `Spatial Transformer` API，可以在该配置下与 PyTorch 实现精度对齐。

该逻辑通过 `condCudnnGridSampler<T>` 进行适配判断，确保仅在支持的设备和数据类型（如 float / double）下调用 cuDNN。

本 PR 对 `grid_sample` 在 GPU 上的实现进行了优化与精度对齐处理。当满足以下特定条件时：
- `padding_mode == zeros`
- `mode == bilinear`
- `align_corners == true`
- `condCudnnGridSampler<T>(x, grid) == true`（即当前环境支持调用 cuDNN 的 grid sampler）

将启用 cuDNN 的 `cudnnSpatialTfSamplerForward` 和 `cudnnSpatialTfSamplerBackward` 接口来计算前向和反向传播，以提高性能并与 PyTorch 在该配置下的行为保持数值一致性。

##### ✅ 前向部分

在 `phi::grid_sample` 前向实现中增加判断逻辑，当满足上述条件时：

- 创建 cuDNN tensor 描述符 `x_desc` 与 `y_desc`
- 创建 cuDNN 空间变换描述符 `st_desc`
- 调用 `cudnnSpatialTfSamplerForward` 实现 bilinear 采样
- 自动释放相关资源（`x_desc` / `y_desc` / `st_desc`）

##### ✅ 反向部分

在反向传播中增加相应逻辑：

- 创建 `x_desc` / `dx_desc` / `y_desc` 描述符
- 创建空间变换描述符 `st_desc`
- 调用 `cudnnSpatialTfSamplerBackward` 计算 `dx` 和 `dgrid`
- 自动释放资源

#### 测试
【前向】：与torch完全对齐。(195个case)
【反向】：存在atomicAdd，在并行计算时，原子操作先后顺序不同导致了结果的随机性。（torch运行两次，都无法对齐）

pcard-67164

### CODE REVIEW

**Review by wanghuancoder — APPROVED** [review]
LGTM

**Review by zrr1999 — APPROVED** [review]
LGTM

File: paddle/phi/kernels/gpu/grid_sample_grad_kernel.cu [hunk 1]
```diff
@@ -612,6 +613,122 @@ void GridSampleGradKernel(const Context& dev_ctx,
...
+    const cudnnDataType_t cudnn_dtype =
+        std::is_same<T, float>::value ? CUDNN_DATA_FLOAT : CUDNN_DATA_DOUBLE;
+
+    PADDLE_ENFORCE_GPU_SUCCESS(
+        phi::dynload::cudnnSetTensor4dDescriptor(x_desc,
+                                                 CUDNN_TENSOR_NCHW,
+                                                 cudnn_dtype,
+                                                 static_cast<int>(N),
+                                                 static_cast<int>(C),
+                                                 static_cast<int>(H_in),
```
Comment (A-nnonymous) [inline paddle/phi/kernels/gpu/grid_sample_grad_kernel.cu:648]: 这里的NCHW，是否涉及到类型收窄？前面声明时采用的可是int64_t，超过2147483647时怎么办？

Comment (A-nnonymous) [reply paddle/phi/kernels/gpu/grid_sample_grad_kernel.cu:648]: 要求加一个溢出判断，或者将声明时的数据类型直接限制为int32_t

**Review by A-nnonymous — CHANGES_REQUESTED** [review]
possible type related issues

File: paddle/phi/kernels/gpu/grid_sample_grad_kernel.cu (hunk 1 above)
Comment (zhengshengning) [reply paddle/phi/kernels/gpu/grid_sample_grad_kernel.cu:648]: 前面 if 判断中使用了 canUse32bitIndexMath 进行了 int32 的限制，超过int32 将不会进入该分支中，所以不会出现溢出的情况。

Comment (A-nnonymous) [reply paddle/phi/kernels/gpu/grid_sample_grad_kernel.cu:648]: LGTM

**Review by A-nnonymous — APPROVED** [review]
